docs/user-guides/acme-http-validation.md

```diff
@@ -0,0 +1,96 @@
+
+The `http01` field simply enables the HTTP challenge for this `Issuer`. No further configuration is necessary or possible.
+
+Once we have created the above `Issuer` we can use it to obtain a certificate.
+
+```yaml
+apiVersion: certmanager.k8s.io/v1alpha1
+kind: Certificate
+metadata:
+  name: nginx-k8s-io
+  namespace: default
+spec:
+  secretName: nginx-k8s-io-tls
+  issuerRef:
+    name: letsencrypt-staging
+  commonName: nginx.k8s.io
+  dnsNames:
+  - nginx2.k8s.io
+  - nginx3.k8s.io
+  acme:
+    config:
+    - http01:
+        ingressClass: nginx
+      domains:
+      - nginx.k8s.io
+    - http01:
+        ingress: my-ingress
+      domains:
+      - nginx2.k8s.io
+      - nginx3.k8s.io
+```
+
+The `Certificate` resource describes our desired certificate and the possible methods that can be used to obtain it. You can learn more about the `Certificate` resource [here][4]. If the certificate is obtained successfully, the resulting key pair will be stored in a secret called `nginx-k8s-io-tls` in the same namespace as the `Certificate`. The certificate will have a common name of `nginx.k8s.io` and the [Subject Alternative Names][6] (SANs) will be `nginx.k8s.io`, `nginx2.k8s.io` and `nginx3.k8s.io`.
+
+In our `Certficate` we have referenced the `letsencrypt-staging` `Issuer` above. The `Issuer` must be in the same namespace as the `Certficate`. If you want to reference a `ClusterIssuer`, which is a cluster-scoped version of an `Issuer`, you must add `kind: ClusterIssuer` to the `issuerRef` stanza.
+
+The `acme` stanza defines the configuration for our ACME challenges. Here we have defined the configuration for our HTTP challenges which will be used to verify domain ownership. For each domain mentioned in an `http01` stanza, `cert-manager` will create a `Pod` that exposes an HTTP endpoint that satisfies the HTTP challenge. For each `http01` stanza, `cert-manager` will create an `Ingress` resource in the same namespace as the `Certificate` with the correct rules to route incoming challenge requests to the `Pods` corresponding to the domains in that stanza. We can control the name of the `Ingress` resource by setting the `ingress` field. If an `Ingress` resource with that name already exists, `cert-manager` will modify it with the correct rules. We can also control the value of the `kubernetes.io/ingress.class` annotation by setting the `ingressClass` field. This will allow us to configure which ingress controller will actually act on these `Ingress` resources. Once domain ownership has been verified, any `cert-manager` affected resources will be cleaned up or deleted. Note that it is your responsibilty to point each domain name at the correct IP address for your ingress controller.
```

Comment:
I'd like to explain the reasoning behind the two options here, and how they interact together. Only one or the other will be used/is required/should be specified.

Basically, if an `ingressName` is specified, that ingress resource *must* already exist, and it will be modified *only* to add the appropriate route for solving the challenge. This is because the GCLB ingress controller, as well as a number of others, create a single load balancer/public IP for each ingress resource, so creating a new ingress resource would create a new load balancer, meaning the challenge would fail.

If the `ingressClass` is specified, a new ingress resource with a randomly generated name will be created in order to solve the challenge, and that resource will then be cleaned up afterwards too (i.e. deleted). This works for the likes of the nginx ingress controller.

If neither are specified, off the top of my head, new ingress resources will be created (like above), but they won't have the ingress class annotation set, so any/all ingress controllers will act upon that resource (depending on cluster configuration).

I *think* if both are set, the ingressName will take precedence and the value of the ingress class field will be ignored.

When cert-manager *modifies* an existing ingress resource (i.e. one specified through ingressName), the *only* thing it will change is adding a new `rule` to solve the challenge.

Reply:
You can see the code path for this here: https://github.com/jetstack-experimental/cert-manager/blob/14cf0d495fbfc6061bf39f5c89f8423a657fec4d/pkg/issuer/acme/http/http.go#L165-L172
You may want to double check over it 😄